### PR TITLE
380807 old dogs

### PR DESCRIPTION
**Pull request by jbarnsley10**

### CODE REVIEW

File: test/unit/app/repos/dogs.test.js [hunk 1]
```diff
@@ -724,4 +725,81 @@ describe('Dog repo', () => {
...
+  })
+
+  describe('getOldDogs', () => {
+    test('should construct correct correct where clause', async () => {
+      sequelize.models.registration.findAll.mockResolvedValue()
+      const statusList = 'Pre-exempt,Exempt'
+
+      await getOldDogs(statusList)
+
+      // console.log('here1', sequelize.models.registration.findAll.mock.calls[0][0].where)
```
Comment (whitewaterdesign) [inline test/unit/app/repos/dogs.test.js:787]: did you want to keep this?

Comment (jbarnsley10) [reply test/unit/app/repos/dogs.test.js:787]: I've redone that test - slight crossover in messages

File: test/unit/app/repos/dogs.test.js [hunk 2]
```diff
@@ -724,4 +725,85 @@ describe('Dog repo', () => {
...
+      await getOldDogs(statusList)
+
+      expect(sequelize.models.registration.findAll).toHaveBeenCalledWith({
+        attributes: expect.anything(),
+        include: expect.anything(),
+        order: expect.anything(),
+        where: expect.anything()
+      })
+    })
+  })
```
Comment (whitewaterdesign) [inline test/unit/app/repos/dogs.test.js:794]: yeah, defo need an api test for this

Comment (jbarnsley10) [reply test/unit/app/repos/dogs.test.js:794]: Yeah, I can't extract the internal detail so this so can't test much. Fortunately, the construction of all the clauses is covered elsewhere

File: app/schema/dogs/get.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
+const Joi = require('joi')
+
+const dogsQueryParamsSchema = Joi.object({
+  forPurging: Joi.string().optional().allow('').allow(null),
+  sortKey: Joi.string().optional().allow('').allow(null),
+  sortOrder: Joi.string().optional().allow('').allow(null)
```
Comment (whitewaterdesign) [inline app/schema/dogs/get.js:6]: what do you think about adding ``.allow('ASC', 'DESC')`` to this?

Comment (whitewaterdesign) [reply app/schema/dogs/get.js:6]: do we want to allow empty and null?

File: app/schema/dogs/get.js [hunk 4]
```diff
@@ -0,0 +1,11 @@
+const Joi = require('joi')
+
+const dogsQueryParamsSchema = Joi.object({
+  forPurging: Joi.string().optional().allow('').allow(null),
```
Comment (whitewaterdesign) [inline app/schema/dogs/get.js:4]: shouldn't this be a boolean?

File: app/routes/dog.js [hunk 5]
```diff
@@ -108,5 +109,33 @@ module.exports = [
...
+      handler: async (request, h) => {
+        let dogs = []
+
+        const sort = {
+          sortKey: request.query.sortKey,
+          sortOrder: request.query.sortOrder
+        }
+
+        if (request.query.forPurging === 'true') {
+          dogs = await getOldDogs('Exempt,Inactive,Withdrawn,Failed', sort) // , new Date(2038, 7, 2))
```
Comment (whitewaterdesign) [inline app/routes/dog.js:132]: take it we want to keep the comment?

File: test/unit/app/repos/dogs.test.js (hunk 2 above)
Comment (whitewaterdesign) [reply test/unit/app/repos/dogs.test.js:794]: suggest we probably need to check that it is mapping the ids to the called withs correctly?

File: app/repos/dogs.js [hunk 6]
```diff
@@ -475,6 +477,100 @@ const deleteDogByIndexNumber = async (indexNumber, user, transaction) => {
...
+  return sortElems.join(',')
+}
+
+const constructDbSort = (options, statusIds) => {
+  const sortDir = options?.sortOrder ?? 'ASC'
+  const order = []
+  const sortKey = options?.sortKey ?? 'status'
+
+  if (sortKey === 'status') {
+    order.push([sequelize.literal(customSort('dog.status_id', statusIds, sortDir)), 'ASC'])
```
Comment (whitewaterdesign) [inline app/repos/dogs.js:493]: great stuff, love this

File: app/routes/dog.js (hunk 5 above)
Comment (jbarnsley10) [reply app/routes/dog.js:132]: Yeah, for now, until I implement a bypass in later code

File: app/repos/dogs.js [hunk 7]
```diff
@@ -475,6 +477,100 @@ const deleteDogByIndexNumber = async (indexNumber, user, transaction) => {
   await sendDeleteToAudit(DOG, dogAggregate, user)
 }
 
+const customSort = (columnName, ids, sortDir) => {
+  const sortElems = []
+  const idList = sortDir === 'ASC' ? ids.reverse() : ids
+  idList.forEach(id => sortElems.push(`${columnName}=${id}`))
+  return sortElems.join(',')
+}
```
Comment (whitewaterdesign) [inline app/repos/dogs.js:485]: v nice addition - could this be pulled out somewhere to make it more generic?

**Review by whitewaterdesign — CHANGES_REQUESTED** [review]
love the sorting approach, works great, just reckon needs a bit more testing coverage and reckon the schema could be stricter.

**Review by whitewaterdesign — APPROVED** [review]
Good stuff

**Pull request merged**